File: hdf-diff/test/modes/baseline-evolution.test.ts
import { describe, it, expect } from 'vitest';
import { diffBaselines, diffHdf } from '../../src/diff.js';
import type { HdfComparison, RequirementDiff } from '../../src/types.js';

// ── Inline fixtures ──────────────────────────────────────────────────

const baselineV1 = {
  name: 'test-stig',
  version: '1.0',
  requirements: [
    {
      id: 'SV-001',
      title: 'Old Title',
      impact: 0.7,
      descriptions: [{ label: 'default', data: 'Check X' }],
      tags: { nist: ['AC-1'] },
    },
    {
      id: 'SV-002',
      title: 'Unchanged',
      impact: 0.5,
      descriptions: [{ label: 'default', data: 'Check Y' }],
      tags: { nist: ['AC-2'] },
    },
    {
      id: 'SV-003',
      title: 'Removed',
      impact: 0.3,
      descriptions: [],
      tags: {},
    },
  ],
};

const baselineV2 = {
  name: 'test-stig',
  version: '2.0',
  requirements: [
    {
      id: 'SV-001',
      title: 'New Title',
      impact: 0.9,
      descriptions: [{ label: 'default', data: 'Check X revised' }],
      tags: { nist: ['AC-1', 'AC-1 (1)'] },
    },
    {
      id: 'SV-002',
      title: 'Unchanged',
      impact: 0.5,
      descriptions: [{ label: 'default', data: 'Check Y' }],
      tags: { nist: ['AC-2'] },
    },
    {
      id: 'SV-004',
      title: 'Added',
      impact: 0.7,
      descriptions: [],
      tags: {},
    },
  ],
};

function findReq(diff: HdfComparison, id: string): RequirementDiff | undefined {
  return diff.requirementDiffs.find((r) => r.id === id);
}

// ── Tests ────────────────────────────────────────────────────────────

describe('baseline evolution comparison mode', () => {
  describe('top-level metadata', () => {
    it('should set comparisonMode to baselineEvolution', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      expect(diff.comparisonMode).toBe('baselineEvolution');
    });

    it('should have formatVersion 1.0.0', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      expect(diff.formatVersion).toBe('1.0.0');
    });

    it('should include a timestamp', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      expect(diff.timestamp).toBeDefined();
      expect(typeof diff.timestamp).toBe('string');
    });
  });

  describe('sources', () => {
    it('should have 2 sources', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      expect(diff.sources).toHaveLength(2);
    });

    it('should label sources with baseline name and version', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      expect(diff.sources[0]!.label).toBe('test-stig 1.0');
      expect(diff.sources[1]!.label).toBe('test-stig 2.0');
    });

    it('should use old/new roles', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      expect(diff.sources[0]!.role).toBe('old');
      expect(diff.sources[1]!.role).toBe('new');
    });
  });

  describe('baselineDiffs', () => {
    it('should show baseline version change', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      expect(diff.baselineDiffs).toHaveLength(1);
      expect(diff.baselineDiffs[0]!.name).toBe('test-stig');
      expect(diff.baselineDiffs[0]!.oldVersion).toBe('1.0');
      expect(diff.baselineDiffs[0]!.newVersion).toBe('2.0');
      expect(diff.baselineDiffs[0]!.state).toBe('updated');
    });

    it('should mark baseline as unchanged when versions match', () => {
      const diff = diffBaselines(baselineV1, baselineV1);
      expect(diff.baselineDiffs[0]!.state).toBe('unchanged');
    });
  });

  describe('requirement states', () => {
    it('should mark SV-001 as updated (title, impact, descriptions, tags changed)', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      const req = findReq(diff, 'SV-001');
      expect(req).toBeDefined();
      expect(req!.state).toBe('updated');
    });

    it('should mark SV-002 as unchanged', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      const req = findReq(diff, 'SV-002');
      expect(req).toBeDefined();
      expect(req!.state).toBe('unchanged');
    });

    it('should mark SV-003 as absent (removed in v2)', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      const req = findReq(diff, 'SV-003');
      expect(req).toBeDefined();
      expect(req!.state).toBe('absent');
      expect(req!.before).not.toBeNull();
      expect(req!.after).toBeNull();
    });

    it('should mark SV-004 as new (added in v2)', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      const req = findReq(diff, 'SV-004');
      expect(req).toBeDefined();
      expect(req!.state).toBe('new');
      expect(req!.before).toBeNull();
      expect(req!.after).not.toBeNull();
    });
  });

  describe('field changes for SV-001', () => {
    it('should have fieldChanges for title, impact, descriptions, and tags', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      const req = findReq(diff, 'SV-001');
      expect(req).toBeDefined();
      expect(req!.fieldChanges.length).toBeGreaterThanOrEqual(4);

      const changedPaths = req!.fieldChanges.map((fc) => fc.path);
      expect(changedPaths).toContain('title');
      expect(changedPaths).toContain('impact');
      expect(changedPaths).toContain('descriptions');
      expect(changedPaths).toContain('tags');
    });

    it('should record title change as replace', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      const req = findReq(diff, 'SV-001');
      const titleChange = req!.fieldChanges.find((fc) => fc.path === 'title');
      expect(titleChange).toBeDefined();
      expect(titleChange!.op).toBe('replace');
      expect(titleChange!.oldValue).toBe('Old Title');
      expect(titleChange!.newValue).toBe('New Title');
    });

    it('should record impact change as replace', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      const req = findReq(diff, 'SV-001');
      const impactChange = req!.fieldChanges.find((fc) => fc.path === 'impact');
      expect(impactChange).toBeDefined();
      expect(impactChange!.op).toBe('replace');
      expect(impactChange!.oldValue).toBe(0.7);
      expect(impactChange!.newValue).toBe(0.9);
    });
  });

  describe('change reasons', () => {
    it('should include impactChanged and metadataChanged for SV-001', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      const req = findReq(diff, 'SV-001');
      expect(req!.changeReasons).toContain('impactChanged');
      expect(req!.changeReasons).toContain('metadataChanged');
    });

    it('should have empty changeReasons for unchanged SV-002', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      const req = findReq(diff, 'SV-002');
      expect(req!.changeReasons).toEqual([]);
    });

    it('should have empty changeReasons for absent SV-003', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      const req = findReq(diff, 'SV-003');
      expect(req!.changeReasons).toEqual([]);
    });

    it('should have empty changeReasons for new SV-004', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      const req = findReq(diff, 'SV-004');
      expect(req!.changeReasons).toEqual([]);
    });
  });

  describe('summary counts', () => {
    it('should have correct totals', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      expect(diff.summary.total).toBe(4);
      expect(diff.summary.matchedCount).toBe(2);
      expect(diff.summary.unmatchedOldCount).toBe(1);
      expect(diff.summary.unmatchedNewCount).toBe(1);
      expect(diff.summary.updated).toBe(1);
      expect(diff.summary.unchanged).toBe(1);
      expect(diff.summary.absent).toBe(1);
      expect(diff.summary.new).toBe(1);
    });

    it('should not have fixed or regressed counts', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      expect(diff.summary.fixed).toBe(0);
      expect(diff.summary.regressed).toBe(0);
    });
  });

  describe('matching config', () => {
    it('should default to exactId strategy', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      expect(diff.matching?.primaryStrategy).toBe('exactId');
    });

    it('should include match metadata on matched requirements', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      const req = findReq(diff, 'SV-001');
      expect(req!.matchStrategy).toBe('exactId');
      expect(req!.matchConfidence).toBe(1.0);
    });
  });

  describe('auto-detection via diffHdf', () => {
    it('should auto-select baselineEvolution mode when both inputs are baselines', () => {
      const diff = diffHdf(baselineV1, baselineV2);
      expect(diff.comparisonMode).toBe('baselineEvolution');
    });

    it('should not auto-select when comparisonMode is explicitly set', () => {
      const diff = diffHdf(baselineV1, baselineV2, { comparisonMode: 'temporal' });
      expect(diff.comparisonMode).toBe('temporal');
    });
  });

  describe('identical baselines', () => {
    it('should classify all requirements as unchanged', () => {
      const diff = diffBaselines(baselineV1, baselineV1);
      for (const req of diff.requirementDiffs) {
        expect(req.state).toBe('unchanged');
      }
    });
  });

  describe('before/after snapshots', () => {
    it('should include full requirement snapshots in before/after', () => {
      const diff = diffBaselines(baselineV1, baselineV2);
      const req = findReq(diff, 'SV-001');
      expect(req!.before).toBeDefined();
      expect(req!.after).toBeDefined();
      expect((req!.before as Record<string, unknown>)['title']).toBe('Old Title');
      expect((req!.after as Record<string, unknown>)['title']).toBe('New Title');
    });
  });
});

File: hdf-diff/go/baseline_evolution_test.go
package diff

import (
	"testing"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
)

// ── Test helpers ─────────────────────────────────────────────────────

func makeBaselineReq(id, title string, impact float64, descs []hdf.Description, tags map[string]any) hdf.BaselineRequirement {
	return hdf.BaselineRequirement{
		ID:           id,
		Title:        strPtr(title),
		Impact:       impact,
		Descriptions: descs,
		Tags:         tags,
		Refs:         []hdf.Reference{},
	}
}

func makeHdfBaseline(name, version string, reqs ...hdf.BaselineRequirement) hdf.HDFBaseline {
	return hdf.HDFBaseline{
		Name:         name,
		Version:      strPtr(version),
		Requirements: reqs,
		Groups:       []hdf.RequirementGroup{},
		Supports:     []hdf.SupportedPlatform{},
		Depends:      []hdf.Dependency{},
		Integrity:    &hdf.Integrity{},
	}
}

// ── Fixtures ─────────────────────────────────────────────────────────

func baselineV1Fixture() hdf.HDFBaseline {
	return makeHdfBaseline("test-stig", "1.0",
		makeBaselineReq("SV-001", "Old Title", 0.7,
			[]hdf.Description{{Label: "default", Data: "Check X"}},
			map[string]any{"nist": []any{"AC-1"}},
		),
		makeBaselineReq("SV-002", "Unchanged", 0.5,
			[]hdf.Description{{Label: "default", Data: "Check Y"}},
			map[string]any{"nist": []any{"AC-2"}},
		),
		makeBaselineReq("SV-003", "Removed", 0.3,
			[]hdf.Description{},
			map[string]any{},
		),
	)
}

func baselineV2Fixture() hdf.HDFBaseline {
	return makeHdfBaseline("test-stig", "2.0",
		makeBaselineReq("SV-001", "New Title", 0.9,
			[]hdf.Description{{Label: "default", Data: "Check X revised"}},
			map[string]any{"nist": []any{"AC-1", "AC-1 (1)"}},
		),
		makeBaselineReq("SV-002", "Unchanged", 0.5,
			[]hdf.Description{{Label: "default", Data: "Check Y"}},
			map[string]any{"nist": []any{"AC-2"}},
		),
		makeBaselineReq("SV-004", "Added", 0.7,
			[]hdf.Description{},
			map[string]any{},
		),
	)
}

// ── Tests ────────────────────────────────────────────────────────────

func TestDiffBaselines_ComparisonMode(t *testing.T) {
	result, err := DiffBaselines(baselineV1Fixture(), baselineV2Fixture(), Options{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.ComparisonMode != ModeBaselineEvolution {
		t.Errorf("expected comparisonMode %q, got %q", ModeBaselineEvolution, result.ComparisonMode)
	}
}

func TestDiffBaselines_FormatVersion(t *testing.T) {
	result, err := DiffBaselines(baselineV1Fixture(), baselineV2Fixture(), Options{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.FormatVersion != "1.0.0" {
		t.Errorf("expected formatVersion %q, got %q", "1.0.0", result.FormatVersion)
	}
}

func TestDiffBaselines_Timestamp(t *testing.T) {
	result, err := DiffBaselines(baselineV1Fixture(), baselineV2Fixture(), Options{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.Timestamp == "" {
		t.Error("expected timestamp to be set")
	}
}

func TestDiffBaselines_Sources(t *testing.T) {
	result, err := DiffBaselines(baselineV1Fixture(), baselineV2Fixture(), Options{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(result.Sources) != 2 {
		t.Fatalf("expected 2 sources, got %d", len(result.Sources))
	}
	if result.Sources[0].Role != RoleOld {
		t.Errorf("expected source[0] role %q, got %q", RoleOld, result.Sources[0].Role)
	}
	if result.Sources[1].Role != RoleNew {
		t.Errorf("expected source[1] role %q, got %q", RoleNew, result.Sources[1].Role)
	}
	if result.Sources[0].Label != "test-stig 1.0" {
		t.Errorf("expected source[0] label %q, got %q", "test-stig 1.0", result.Sources[0].Label)
	}
	if result.Sources[1].Label != "test-stig 2.0" {
		t.Errorf("expected source[1] label %q, got %q", "test-stig 2.0", result.Sources[1].Label)
	}
}

func TestDiffBaselines_BaselineDiffs(t *testing.T) {
	result, err := DiffBaselines(baselineV1Fixture(), baselineV2Fixture(), Options{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(result.BaselineDiffs) != 1 {
		t.Fatalf("expected 1 baseline diff, got %d", len(result.BaselineDiffs))
	}
	bd := result.BaselineDiffs[0]
	if bd.Name != "test-stig" {
		t.Errorf("expected baseline name %q, got %q", "test-stig", bd.Name)
	}
	if bd.OldVersion != "1.0" {
		t.Errorf("expected oldVersion %q, got %q", "1.0", bd.OldVersion)
	}
	if bd.NewVersion != "2.0" {
		t.Errorf("expected newVersion %q, got %q", "2.0", bd.NewVersion)
	}
	if bd.State != StateUpdated {
		t.Errorf("expected baseline state %q, got %q", StateUpdated, bd.State)
	}
}

func TestDiffBaselines_BaselineDiffsUnchanged(t *testing.T) {
	v1 := baselineV1Fixture()
	result, err := DiffBaselines(v1, v1, Options{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(result.BaselineDiffs) != 1 {
		t.Fatalf("expected 1 baseline diff, got %d", len(result.BaselineDiffs))
	}
	if result.BaselineDiffs[0].State != StateUnchanged {
		t.Errorf("expected baseline state %q, got %q", StateUnchanged, result.BaselineDiffs[0].State)
	}
}

func TestDiffBaselines_SV001Updated(t *testing.T) {
	result, err := DiffBaselines(baselineV1Fixture(), baselineV2Fixture(), Options{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	req := findReq(result.RequirementDiffs, "SV-001")
	if req == nil {
		t.Fatal("SV-001 not found")
	}
	if req.State != StateUpdated {
		t.Errorf("expected SV-001 state %q, got %q", StateUpdated, req.State)
	}
}

func TestDiffBaselines_SV002Unchanged(t *testing.T) {
	result, err := DiffBaselines(baselineV1Fixture(), baselineV2Fixture(), Options{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	req := findReq(result.RequirementDiffs, "SV-002")
	if req == nil {
		t.Fatal("SV-002 not found")
	}
	if req.State != StateUnchanged {
		t.Errorf("expected SV-002 state %q, got %q", StateUnchanged, req.State)
	}
}

func TestDiffBaselines_AbsentAndNew(t *testing.T) {
	result, err := DiffBaselines(baselineV1Fixture(), baselineV2Fixture(), Options{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	tests := []struct {
		id            string
		expectedState RequirementState
		beforeNil     bool
		afterNil      bool
	}{
		{"SV-003", StateAbsent, false, true},
		{"SV-004", StateNew, true, false},
	}

	for _, tc := range tests {
		t.Run(tc.id, func(t *testing.T) {
			req := findReq(result.RequirementDiffs, tc.id)
			if req == nil {
				t.Fatalf("%s not found", tc.id)
			}
			if req.State != tc.expectedState {
				t.Errorf("expected state %q, got %q", tc.expectedState, req.State)
			}
			if tc.beforeNil && req.Before != nil {
				t.Error("expected before to be nil")
			}
			if !tc.beforeNil && req.Before == nil {
				t.Error("expected before to be non-nil")
			}
			if tc.afterNil && req.After != nil {
				t.Error("expected after to be nil")
			}
			if !tc.afterNil && req.After == nil {
				t.Error("expected after to be non-nil")
			}
		})
	}
}

func TestDiffBaselines_FieldChanges(t *testing.T) {
	result, err := DiffBaselines(baselineV1Fixture(), baselineV2Fixture(), Options{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	req := findReq(result.RequirementDiffs, "SV-001")
	if req == nil {
		t.Fatal("SV-001 not found")
	}
	if len(req.FieldChanges) < 4 {
		t.Errorf("expected at least 4 field changes, got %d", len(req.FieldChanges))
	}

	changedPaths := make(map[string]bool)
	for _, fc := range req.FieldChanges {
		changedPaths[fc.Path] = true
	}

	for _, expectedPath := range []string{"title", "impact", "descriptions", "tags"} {
		if !changedPaths[expectedPath] {
			t.Errorf("expected field change for %q, not found", expectedPath)
		}
	}
}

func TestDiffBaselines_ChangeReasons(t *testing.T) {
	result, err := DiffBaselines(baselineV1Fixture(), baselineV2Fixture(), Options{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	// SV-001: should have impactChanged and metadataChanged
	req := findReq(result.RequirementDiffs, "SV-001")
	if req == nil {
		t.Fatal("SV-001 not found")
	}
	hasImpactChanged := false
	hasMetadataChanged := false
	for _, r := range req.ChangeReasons {
		if r == ReasonImpactChanged {
			hasImpactChanged = true
		}
		if r == ReasonMetadataChanged {
			hasMetadataChanged = true
		}
	}
	if !hasImpactChanged {
		t.Error("expected impactChanged in SV-001 change reasons")
	}
	if !hasMetadataChanged {
		t.Error("expected metadataChanged in SV-001 change reasons")
	}

	// SV-002: should have no change reasons
	req2 := findReq(result.RequirementDiffs, "SV-002")
	if req2 == nil {
		t.Fatal("SV-002 not found")
	}
	if len(req2.ChangeReasons) != 0 {
		t.Errorf("expected 0 change reasons for SV-002, got %d", len(req2.ChangeReasons))
	}
}

func TestDiffBaselines_SummaryCounts(t *testing.T) {
	result, err := DiffBaselines(baselineV1Fixture(), baselineV2Fixture(), Options{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	s := result.Summary
	if s.Total != 4 {
		t.Errorf("expected total 4, got %d", s.Total)
	}
	if s.MatchedCount != 2 {
		t.Errorf("expected matchedCount 2, got %d", s.MatchedCount)
	}
	if s.UnmatchedOldCount != 1 {
		t.Errorf("expected unmatchedOldCount 1, got %d", s.UnmatchedOldCount)
	}
	if s.UnmatchedNewCount != 1 {
		t.Errorf("expected unmatchedNewCount 1, got %d", s.UnmatchedNewCount)
	}
	if s.Updated != 1 {
		t.Errorf("expected updated 1, got %d", s.Updated)
	}
	if s.Unchanged != 1 {
		t.Errorf("expected unchanged 1, got %d", s.Unchanged)
	}
	if s.Absent != 1 {
		t.Errorf("expected absent 1, got %d", s.Absent)
	}
	if s.New != 1 {
		t.Errorf("expected new 1, got %d", s.New)
	}
	if s.Fixed != 0 {
		t.Errorf("expected fixed 0, got %d", s.Fixed)
	}
	if s.Regressed != 0 {
		t.Errorf("expected regressed 0, got %d", s.Regressed)
	}
}

func TestDiffBaselines_MatchingConfig(t *testing.T) {
	result, err := DiffBaselines(baselineV1Fixture(), baselineV2Fixture(), Options{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.Matching == nil {
		t.Fatal("expected matching config to be set")
	}
	if result.Matching.PrimaryStrategy != "exactId" {
		t.Errorf("expected primaryStrategy %q, got %q", "exactId", result.Matching.PrimaryStrategy)
	}

	req := findReq(result.RequirementDiffs, "SV-001")
	if req == nil {
		t.Fatal("SV-001 not found")
	}
	if req.MatchStrategy != "exactId" {
		t.Errorf("expected matchStrategy %q, got %q", "exactId", req.MatchStrategy)
	}
	if req.MatchConfidence == nil || *req.MatchConfidence != 1.0 {
		t.Error("expected matchConfidence 1.0")
	}
}

func TestDiffBaselines_IdenticalBaselines(t *testing.T) {
	v1 := baselineV1Fixture()
	result, err := DiffBaselines(v1, v1, Options{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	for _, req := range result.RequirementDiffs {
		if req.State != StateUnchanged {
			t.Errorf("expected all requirements unchanged, got %q for %s", req.State, req.ID)
		}
	}
}

func TestDiffBaselines_BeforeAfterSnapshots(t *testing.T) {
	result, err := DiffBaselines(baselineV1Fixture(), baselineV2Fixture(), Options{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	req := findReq(result.RequirementDiffs, "SV-001")
	if req == nil {
		t.Fatal("SV-001 not found")
	}
	if req.Before == nil {
		t.Fatal("expected before snapshot")
	}
	if req.After == nil {
		t.Fatal("expected after snapshot")
	}
	if req.Before.Title == nil || *req.Before.Title != "Old Title" {
		t.Error("expected before title 'Old Title'")
	}
	if req.After.Title == nil || *req.After.Title != "New Title" {
		t.Error("expected after title 'New Title'")
	}
}
